File: app/src/main/java/com/company/art_and_culture/myarts/museum_fragment/MuseumInfoDataSource.java
package com.company.art_and_culture.myarts.museum_fragment;

import android.app.Application;

import com.company.art_and_culture.myarts.Constants;
import com.company.art_and_culture.myarts.network.NetworkQuery;
import com.company.art_and_culture.myarts.pojo.ArtProvider;
import com.company.art_and_culture.myarts.pojo.Maker;
import com.company.art_and_culture.myarts.pojo.ServerRequest;
import com.company.art_and_culture.myarts.pojo.ServerResponse;

import java.util.ArrayList;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class MuseumInfoDataSource {

    private Application application;
    private String artProviderId;
    private MutableLiveData<ArtProvider> artProvider = new MutableLiveData<>();
    private MutableLiveData<ArrayList<Maker>> listMakers = new MutableLiveData<>();
    private MutableLiveData<Boolean> isLiked = new MutableLiveData<>();
    private MutableLiveData<Boolean> isLoading = new MutableLiveData<>();

    public MuseumInfoDataSource(Application application, String museumId) {
        this.artProviderId = museumId;
        this.application = application;
        updateIsLoadingState(true);
        loadMuseumInfo();
        loadMakersList();
    }

    public void setArtProviderId(String artProviderId) {
        this.artProviderId = artProviderId;
        updateIsLoadingState(true);
        loadMuseumInfo();
        loadMakersList();
    }

    public void refresh() {
        updateIsLoadingState(true);
        loadMuseumInfo();
        loadMakersList();
    }

    public LiveData<ArtProvider> getArtProvider() {
        return artProvider;
    }

    public void updateArtProvider(ArtProvider artProvider) {
        this.artProvider.postValue(artProvider);
    }

    public LiveData<ArrayList<Maker>> getListMakers() {
        return listMakers;
    }

    private void updateMakersList(ArrayList<Maker> listMakers) {
        this.listMakers.postValue(listMakers);
    }

    private void updateArtProviderLike(boolean liked) {
        isLiked.postValue(liked);
    }

    public LiveData<Boolean> getArtProviderLike() {
        return isLiked;
    }

    private void updateIsLoadingState(Boolean state) {
        isLoading.postValue(state);
    }

    public LiveData<Boolean> getIsLoading() {
        return isLoading;
    }

    private void loadMuseumInfo() {

        String userUniqueId = application.getSharedPreferences(Constants.TAG,0).getString(Constants.USER_UNIQUE_ID,"");
        ServerRequest request = new ServerRequest();
        request.setOperation(Constants.GET_MUSEUM_INFO_OPERATION);
        request.setUserUniqueId(userUniqueId);
        request.setArtProviderId(artProviderId);

        Call<ServerResponse> response = NetworkQuery.getInstance().create(Constants.BASE_URL, request);
        response.enqueue(new Callback<ServerResponse>() {
            @Override
            public void onResponse(Call<ServerResponse> call, Response<ServerResponse> response) {

                updateIsLoadingState(false);

                if (response.isSuccessful()) {
                    ServerResponse resp = response.body();
                    if(resp.getResult().equals(Constants.SUCCESS)) {

                        updateArtProvider(resp.getArtProvider());

                    } else { }
                } else { }
            }

            @Override
            public void onFailure(Call<ServerResponse> call, Throwable t) {
                updateIsLoadingState(false);
            }
        });
    }

    private void loadMakersList() {

        String userUniqueId = application.getSharedPreferences(Constants.TAG,0).getString(Constants.USER_UNIQUE_ID,"");
        ServerRequest request = new ServerRequest();
        request.setOperation(Constants.GET_MUSEUM_MAKERS_LIST_OPERATION);
        request.setUserUniqueId(userUniqueId);
        request.setArtProviderId(artProviderId);

        Call<ServerResponse> response = NetworkQuery.getInstance().create(Constants.BASE_URL, request);
        response.enqueue(new Callback<ServerResponse>() {
            @Override
            public void onResponse(Call<ServerResponse> call, Response<ServerResponse> response) {

                if (response.isSuccessful()) {
                    ServerResponse resp = response.body();
                    if(resp.getResult().equals(Constants.SUCCESS)) {

                        updateMakersList(resp.getListMakers());

                    } else { }
                } else { }
            }
            @Override
            public void onFailure(Call<ServerResponse> call, Throwable t) { }
        });
    }

    public void likeMuseum(ArtProvider museum, String userUniqueId) {
        ServerRequest request = new ServerRequest();
        request.setUserUniqueId(userUniqueId);
        request.setOperation(Constants.MUSEUM_LIKE_OPERATION);
        request.setArtProvider(museum);
        Call<ServerResponse> response = NetworkQuery.getInstance().create(Constants.BASE_URL, request);
        response.enqueue(new Callback<ServerResponse>() {
            @Override
            public void onResponse(Call<ServerResponse> call, Response<ServerResponse> response) {
                if (response.isSuccessful()) {
                    ServerResponse resp = response.body();
                    if(resp.getResult().equals(Constants.SUCCESS)) {

                        updateArtProviderLike(resp.getArtProvider().isLiked());

                    } else { }
                } else { }
            }
            @Override
            public void onFailure(Call<ServerResponse> call, Throwable t) { }
        });
    }
}

File: app/src/main/java/com/company/art_and_culture/myarts/museum_fragment/MuseumInfoDataSource.kt
package com.company.art_and_culture.myarts.museum_fragment

import android.util.Log
import androidx.lifecycle.MutableLiveData
import androidx.paging.PagingSource
import com.company.art_and_culture.myarts.Constants
import com.company.art_and_culture.myarts.network.MainHttpService
import com.company.art_and_culture.myarts.pojo.*
import kotlinx.coroutines.runBlocking

class MuseumInfoDataSource(
    private val mainHttpService: MainHttpService,
    private var artProviderId: String,
    private val userUniqueId: String
) {

    var artProvider = MutableLiveData<ArtProvider?>()
        private set
    var listMakers = MutableLiveData<ArrayList<Maker>?>()
        private set
    var isLoading = MutableLiveData<Boolean>()
        private set
    var isArtProviderLiked = MutableLiveData<Boolean?>()
        private set
    var isContentEmpty = MutableLiveData<Boolean>()
        private set

    fun setMuseumId(artProviderId: String) {
        isLoading.postValue(true)
        this.artProviderId = artProviderId

        runBlocking {
            loadMuseumInfo()
            loadMakersList()
        }
    }

    fun refresh() {
        isLoading.postValue(true)

        runBlocking {
            loadMuseumInfo()
            loadMakersList()
        }
    }

    private suspend fun loadMuseumInfo() {

        val request = ServerRequest()
        request.setOperation(Constants.GET_MUSEUM_INFO_OPERATION)
        request.setUserUniqueId(userUniqueId)
        request.setArtProviderId(artProviderId)

        try {
            val response = mainHttpService.mainRequest(request)

            if (response.isSuccessful) {
                val serverResponse = response.body()
                if (serverResponse?.result == Constants.SUCCESS) {

                    isLoading.postValue(false)
                    isContentEmpty.postValue(false)
                    artProvider.postValue(serverResponse.artProvider)

                } else { }
            } else { }
        } catch (e: Exception) {
            isLoading.postValue(false)
        }
    }

    private suspend fun loadMakersList() {

        val request = ServerRequest()
        request.setOperation(Constants.GET_MUSEUM_MAKERS_LIST_OPERATION)
        request.setUserUniqueId(userUniqueId)
        request.setArtProviderId(artProviderId)

        try {
            val response = mainHttpService.mainRequest(request)
            if (response.isSuccessful) {
                val serverResponse = response.body()
                if (serverResponse?.result == Constants.SUCCESS) {

                    listMakers.postValue(serverResponse.listMakers)

                } else { }

            } else { }
        } catch (e: Exception) { }
    }

    suspend fun likeMuseum(museum: ArtProvider?, userUniqueId: String?) {

        val request = ServerRequest()
        request.setUserUniqueId(userUniqueId)
        request.setOperation(Constants.MUSEUM_LIKE_OPERATION)
        request.setArtProvider(museum)

        try {
            val response = mainHttpService.mainRequest(request)
            if (response.isSuccessful) {
                val serverResponse = response.body()
                if (serverResponse?.result == Constants.SUCCESS) {

                    isArtProviderLiked.postValue(serverResponse.artProvider.isLiked)

                } else { }
            } else { }

        } catch (e: Exception) { }
    }

    suspend fun writeDimensionsOnServer(art: Art?) {
        val request = ServerRequest()
        request.setOperation(Constants.ART_WRITE_DIMENS_OPERATION)
        request.setArt(art)

        try {
            val response = mainHttpService.mainRequest(request)
        } catch (e: Exception) { }

    }

    fun onDestroy() {
        artProvider.postValue(null)
        isArtProviderLiked.postValue(null)
        listMakers.postValue(null)
        isLoading.postValue(true)
        isContentEmpty.postValue(true)
    }


}
